src/shared/selinux-util.c

```diff
@@ -434,40 +444,84 @@ int mac_selinux_get_child_mls_label(int socket_fd, const char *exe, const char *
         if (r < 0)
                 return -errno;
 
-        if (!exec_label) {
+        /* If SELinuxContext= was set in the unit, use it */
+        if (exec_label) {
+                targetcon = strdup(exec_label);
+                if (!targetcon)
+                        return -ENOMEM;
+        } else {
                 /* If there is no context set for next exec let's use context
                    of target executable */
-                r = getfilecon_raw(exe, &fcon);
+                r = getfilecon_raw(exe, &targetcon);
                 if (r < 0)
                         return -errno;
         }
 
-        bcon = context_new(mycon);
-        if (!bcon)
+        /* If neither SELinuxContext= was set,
+         * nor a SELinux label exists on the executable file,
+         * use the context under which systemd is running */
+        if (!targetcon) {
```

Comment:
hmm, with your patch targetcon is always non-NULL, so why this check? what am I missing?

Reply:
If exec_label is not-null, targetcon is set. Otherwise `getfilecon_raw(exe, &targetcon)` is called, but targetcon will remain null if there is no SELinux label on exe. I should probably add a comment explaining this?

Here is a small test program to make sure that it will remain null if getfilecon_raw() does not find a label:

```c
#include <selinux/selinux.h>
#include <stdio.h>
int main()
{
	char *context = NULL;
	getfilecon_raw("/tmp/t.c", &context);
	if (context) {
		printf("Context is not null\n");
	} else {
		printf("Context is null\n");
	}
}
```

I've saved it as /tmp/t.c and run it on a SELinux and a non-SELinux systems and it works as expected, printing "Context is not null" on a SELinux-enabled system and "Context is null" on a non-SELinux-enabled system.

Reply:
hm, there is already a comment: "nor a SELinux label exists on the executable file"

Reply:
getfilecon_raw() returns ENODATA in case no context is set. Look for that.

Reply:
Oh, yes, this code is incorrect because "return -errno" will happen if there is no context set